File: sdks/go/mcp/engine.go
package mcp

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strings"
	"sync"

	solvapay "github.com/solvapay/solvapay-go"
)

// HTTPConfig is the public origin / product settings for [Engine].
type HTTPConfig struct {
	ProductRef    string
	PublicBaseURL string
	ResourceURI   string
	MCPPath       string
	Views         []string
	OauthPaths    map[string]any
}

// Engine routes OAuth through McpOauthRequest and /mcp through McpDispatch.
type Engine struct {
	client   *solvapay.Client
	cfg      HTTPConfig
	mu       sync.Mutex
	payables map[string]Options
}

// NewEngine binds a SolvaPay client to one MCP origin.
func NewEngine(client *solvapay.Client, cfg HTTPConfig) *Engine {
	if cfg.MCPPath == "" {
		cfg.MCPPath = "/mcp"
	}
	if cfg.ResourceURI == "" {
		cfg.ResourceURI = "ui://widget.html"
	}
	return &Engine{client: client, cfg: cfg, payables: map[string]Options{}}
}

// RegisterPayable records a merchant tool. Names feed EngineConfig.payableTools.
func (e *Engine) RegisterPayable(name string, opts Options) error {
	if name == "" {
		return fmt.Errorf("tool name is required")
	}
	if opts.Client == nil {
		opts.Client = e.client
	}
	if opts.Product == "" {
		return fmt.Errorf("Product is required")
	}
	if opts.Handler == nil {
		return fmt.Errorf("Handler is required")
	}
	if opts.UsageType == "" {
		opts.UsageType = "requests"
	}
	e.mu.Lock()
	defer e.mu.Unlock()
	e.payables[name] = opts
	return nil
}

// ServeHTTP implements net/http.Handler.
func (e *Engine) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	path := r.URL.Path
	if r.URL.RawQuery != "" {
		path = path + "?" + r.URL.RawQuery
	}
	if pathOnly(r.URL.Path) != e.cfg.MCPPath {
		e.handleOAuth(w, r, path)
		return
	}
	if r.Method != http.MethodPost {
		w.Header().Set("Allow", "POST")
		w.WriteHeader(http.StatusMethodNotAllowed)
		return
	}
	e.handleMCP(w, r)
}

func (e *Engine) handleOAuth(w http.ResponseWriter, r *http.Request, path string) {
	body, err := io.ReadAll(r.Body)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadRequest)
		return
	}
	headers := map[string]string{}
	for key, values := range r.Header {
		if len(values) > 0 {
			headers[strings.ToLower(key)] = values[0]
		}
	}
	envelope, err := e.client.McpOauthRequest(r.Context(), map[string]any{
		"method":  r.Method,
		"path":    path,
		"headers": headers,
		"body":    string(body),
		"config": map[string]any{
			"publicBaseUrl": e.cfg.PublicBaseURL,
			"mcpPath":       e.cfg.MCPPath,
			"productRef":    e.cfg.ProductRef,
			"oauthPaths":    e.cfg.OauthPaths,
		},
	})
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	writeOAuth(w, envelope)
}

func (e *Engine) handleMCP(w http.ResponseWriter, r *http.Request) {
	var rpc any
	if err := json.NewDecoder(r.Body).Decode(&rpc); err != nil {
		http.Error(w, "invalid JSON-RPC body", http.StatusBadRequest)
		return
	}
	e.mu.Lock()
	names := make([]string, 0, len(e.payables))
	for name := range e.payables {
		names = append(names, name)
	}
	e.mu.Unlock()
	params := map[string]any{
		"rpc": rpc,
		"config": map[string]any{
			"productRef":    e.cfg.ProductRef,
			"publicBaseUrl": e.cfg.PublicBaseURL,
			"resourceUri":   e.cfg.ResourceURI,
			"payableTools":  names,
			"mcpPath":       e.cfg.MCPPath,
			"views":         e.cfg.Views,
			"userAgent":     r.Header.Get("User-Agent"),
		},
	}
	if auth := r.Header.Get("Authorization"); auth != "" {
		params["authHeader"] = auth
	}
	envelope, err := e.client.McpDispatch(r.Context(), params)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	kind, _ := asString(asMap(envelope)["kind"])
	switch kind {
	case "rpc":
		writeJSON(w, http.StatusOK, asMap(envelope)["rpc"])
	case "challenge":
		env := asMap(envelope)
		status := asInt(env["status"], 401)
		for k, v := range asStringMap(env["headers"]) {
			w.Header().Set(k, v)
		}
		writeJSON(w, status, env["body"])
	case "invokeHandler":
		e.resumePayable(w, r, asMap(envelope))
	default:
		http.Error(w, fmt.Sprintf("unexpected mcpDispatch kind: %s", kind), http.StatusInternalServerError)
	}
}

func (e *Engine) resumePayable(w http.ResponseWriter, r *http.Request, envelope map[string]any) {
	tool, _ := asString(envelope["tool"])
	token, _ := asString(envelope["token"])
	e.mu.Lock()
	opts, ok := e.payables[tool]
	e.mu.Unlock()
	if !ok {
		http.Error(w, "unknown payable tool: "+tool, http.StatusInternalServerError)
		return
	}
	args := asMap(envelope["args"])
	if _, has := args["customer_ref"]; !has {
		if ref, ok := asString(envelope["customerRef"]); ok && ref != "" {
			args["customer_ref"] = ref
		}
	}
	result, err := InvokePayable(r.Context(), args, opts)
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	handlerEnvelope, err := json.Marshal(result)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	var handler any
	if err := json.Unmarshal(handlerEnvelope, &handler); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	resumed, err := CallSync(r.Context(), "mcpResume", map[string]any{
		"token":           token,
		"handlerEnvelope": handler,
	})
	if err != nil {
		http.Error(w, err.Error(), http.StatusBadGateway)
		return
	}
	var parsed any
	if err := json.Unmarshal(resumed, &parsed); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}
	writeJSON(w, http.StatusOK, asMap(parsed)["rpc"])
}

func pathOnly(path string) string {
	if i := strings.IndexByte(path, '?'); i >= 0 {
		return path[:i]
	}
	return path
}

func writeOAuth(w http.ResponseWriter, envelope any) {
	env := asMap(envelope)
	status := asInt(env["status"], 500)
	for k, v := range asStringMap(env["headers"]) {
		w.Header().Set(k, v)
	}
	writeJSON(w, status, env["body"])
}

func writeJSON(w http.ResponseWriter, status int, body any) {
	if body == nil {
		w.WriteHeader(status)
		return
	}
	if _, ok := w.Header()["Content-Type"]; !ok {
		w.Header().Set("Content-Type", "application/json")
	}
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func asMap(v any) map[string]any {
	if m, ok := v.(map[string]any); ok {
		if m == nil {
			return map[string]any{}
		}
		return m
	}
	return map[string]any{}
}

func asStringMap(v any) map[string]string {
	out := map[string]string{}
	for k, val := range asMap(v) {
		if s, ok := val.(string); ok {
			out[k] = s
		}
	}
	return out
}

func asString(v any) (string, bool) {
	s, ok := v.(string)
	return s, ok
}

func asInt(v any, fallback int) int {
	switch n := v.(type) {
	case float64:
		return int(n)
	case int:
		return n
	case json.Number:
		i, _ := n.Int64()
		return int(i)
	default:
		return fallback
	}
}

File: sdks/ruby-mcp/lib/solvapay/mcp/engine.rb
# frozen_string_literal: true

require "json"
require "solvapay"

module SolvaPay
  module Mcp
    # Rack adapter: OAuth via mcp_oauth_request, /mcp via mcp_dispatch.
    class Engine
      def initialize(client:, product_ref:, public_base_url:, resource_uri: "ui://widget.html", mcp_path: "/mcp",
                     views: nil, oauth_paths: nil, hs256_secret: nil, jwks_json: nil)
        raise ArgumentError, "client is required" if client.nil?
        raise ArgumentError, "product_ref is required" if product_ref.nil? || product_ref.empty?
        raise ArgumentError, "public_base_url is required" if public_base_url.nil? || public_base_url.empty?

        @client = client
        @facade = SolvaPay::Facade.new(api_client: client)
        @product_ref = product_ref
        @public_base_url = public_base_url
        @resource_uri = resource_uri.nil? || resource_uri.empty? ? "ui://widget.html" : resource_uri
        @mcp_path = mcp_path.nil? || mcp_path.empty? ? "/mcp" : mcp_path
        @views = views
        @oauth_paths = oauth_paths
        @hs256_secret = hs256_secret
        @jwks_json = jwks_json
        @payables = {} #: Hash[String, untyped]
        @mutex = Mutex.new
      end

      def register_payable(name, product:, handler:, get_customer_ref: nil, usage_type: "requests",
                           title: nil, description: nil, input_schema: nil, annotations: nil)
        raise ArgumentError, "tool name is required" if name.nil? || name.empty?
        raise ArgumentError, "product is required" if product.nil? || product.empty?
        raise ArgumentError, "handler is required" if handler.nil?

        @mutex.synchronize do
          @payables[name] = {
            product: product,
            handler: handler,
            get_customer_ref: get_customer_ref,
            usage_type: usage_type.nil? || usage_type.empty? ? "requests" : usage_type,
            title: title,
            description: description,
            input_schema: input_schema,
            annotations: annotations,
          }
        end
        self
      end

      def call(env)
        path = env["PATH_INFO"].to_s
        query = env["QUERY_STRING"].to_s
        full_path = query.empty? ? path : "#{path}?#{query}"
        method = env["REQUEST_METHOD"].to_s
        return handle_oauth(env, method, full_path) unless path == @mcp_path
        return [405, { "allow" => "POST, OPTIONS" }, []] unless method == "POST"

        handle_mcp(env)
      end

      private

      def handle_oauth(env, method, path)
        envelope = @client.mcp_oauth_request(
          params: {
            "method" => method,
            "path" => path,
            "headers" => rack_headers(env),
            "body" => read_body(env),
            "config" => oauth_config,
          },
        )
        oauth_response(envelope)
      end

      def handle_mcp(env)
        rpc = nil
        begin
          rpc = JSON.parse(read_body(env))
          html = SolvaPay::Mcp.widget_html_rpc(rpc, @resource_uri, @public_base_url, @product_ref, @views)
          return json_response(200, html) unless html.nil?

          payable_tools = @mutex.synchronize { payable_tool_specs }
          params = {
            "rpc" => rpc,
            "config" => {
              "productRef" => @product_ref,
              "publicBaseUrl" => @public_base_url,
              "resourceUri" => @resource_uri,
              "payableTools" => payable_tools,
              "mcpPath" => @mcp_path,
              "views" => @views,
              "userAgent" => env["HTTP_USER_AGENT"],
            },
          }
          params["config"]["hs256Secret"] = @hs256_secret unless @hs256_secret.nil? || @hs256_secret.empty?
          params["config"]["jwksJson"] = @jwks_json unless @jwks_json.nil?
          auth = env["HTTP_AUTHORIZATION"]
          params["authHeader"] = auth unless auth.nil? || auth.empty?
          proto = env["HTTP_MCP_PROTOCOL_VERSION"]
          params["mcpProtocolVersionHeader"] = proto unless proto.nil? || proto.empty?
          envelope = stringify(@client.mcp_dispatch(params: params))
          case envelope["kind"]
          when "rpc"
            status = envelope["status"]
            json_response(status.is_a?(Integer) ? status : 200, envelope["rpc"])
          when "challenge"
            headers = stringify_headers(envelope["headers"])
            merge_native_cors(env, headers)
            [Integer(envelope["status"] || 401), headers, [encode_body(envelope["body"])]]
          when "invokeHandler"
            resume_payable(envelope)
          else
            raise SolvaPay::SolvaPayError.new("unexpected mcpDispatch kind: #{envelope['kind'].inspect}",
                                              code: "invalid_dispatch",)
          end
        rescue JSON::ParserError
          json_rpc_error(400, nil, -32_700, "Parse error")
        rescue StandardError => e
          warn e.full_message
          id = rpc.is_a?(Hash) ? rpc["id"] : nil
          json_rpc_error(200, id, -32_603, e.message)
        end
      end

      def resume_payable(envelope)
        tool = envelope["tool"].to_s
        spec = @mutex.synchronize { @payables[tool] }
        raise SolvaPay::SolvaPayError.new("unknown payable tool: #{tool}", code: "unknown_tool") if spec.nil?

        empty_args = {} #: Hash[untyped, untyped]
        args = envelope["args"].is_a?(Hash) ? stringify(envelope["args"]) : empty_args
        args = symbolize(args)
        unless args.key?(:customer_ref)
          ref = envelope["customerRef"]
          args[:customer_ref] = ref if ref.is_a?(String) && !ref.empty?
        end
        result = SolvaPay::Mcp.invoke_payable(
          solvapay: @facade,
          product: spec[:product],
          handler: spec[:handler],
          get_customer_ref: spec[:get_customer_ref],
          usage_type: spec[:usage_type],
          args: args,
        )
        resumed = stringify(
          SolvaPay::Mcp::Core.call(
            "mcpResume",
            { "token" => envelope["token"], "handlerEnvelope" => result },
          ),
        )
        json_response(200, resumed["rpc"] || resumed)
      end

      def payable_tool_specs
        @payables.keys.sort.map do |name|
          spec = @payables.fetch(name)
          entry = { "name" => name }
          entry["title"] = spec[:title] unless spec[:title].nil?
          entry["description"] = spec[:description] unless spec[:description].nil?
          entry["inputSchema"] = spec[:input_schema] unless spec[:input_schema].nil?
          entry["annotations"] = spec[:annotations] unless spec[:annotations].nil?
          entry
        end
      end

      def oauth_config
        config = {
          "publicBaseUrl" => @public_base_url,
          "mcpPath" => @mcp_path,
          "productRef" => @product_ref,
        }
        config["oauthPaths"] = @oauth_paths unless @oauth_paths.nil?
        config
      end

      def oauth_response(envelope)
        envelope = stringify(envelope)
        status = Integer(envelope["status"] || 500)
        headers = stringify_headers(envelope["headers"])
        [status, headers, [encode_body(envelope["body"])]]
      end

      def json_rpc_error(status, id, code, message)
        json_response(status, { "jsonrpc" => "2.0", "id" => id, "error" => { "code" => code, "message" => message } })
      end

      def json_response(status, body)
        [
          status,
          { "content-type" => "application/json" },
          [encode_body(body)],
        ]
      end

      def encode_body(body)
        return "" if body.nil?
        return body if body.is_a?(String)

        JSON.generate(body)
      end

      def merge_native_cors(env, headers)
        origin = env["HTTP_ORIGIN"]
        result = stringify(SolvaPay::Mcp::Core.call("mcpNativeCors", { "origin" => origin }))
        cors = result["headers"]
        return unless cors.is_a?(Hash)

        cors.each do |key, value|
          headers[key.to_s.downcase] = value.to_s if value.is_a?(String)
        end
      end

      def stringify_headers(headers)
        out = {} #: Hash[String, String]
        return out unless headers.is_a?(Hash)

        headers.each do |key, value|
          out[key.to_s.downcase] = value.to_s if value.is_a?(String)
        end
        out
      end

      def rack_headers(env)
        headers = {} #: Hash[String, String]
        env.each do |key, value|
          next unless key.start_with?("HTTP_") && value.is_a?(String)

          name = key.delete_prefix("HTTP_").downcase.tr("_", "-")
          headers[name] = value
        end
        headers["content-type"] = env["CONTENT_TYPE"] if env["CONTENT_TYPE"].is_a?(String)
        headers
      end

      def read_body(env)
        input = env["rack.input"]
        return "" if input.nil?

        input.rewind if input.respond_to?(:rewind)
        input.read.to_s
      end

      def stringify(value)
        JSON.parse(JSON.generate(value))
      end

      def symbolize(hash)
        acc = {} #: Hash[Symbol, untyped]
        hash.each do |key, value|
          acc[key.to_sym] = value
        end
        acc
      end
    end
  end
end
